Add end location to all applicable elements

arogachev opened the pull request:
A feature extracted from #224 / #225.

Use case - having links like this - https://github.com/yiisoft/yii2/blob/e459214a927424bafd7a23ed946951d4a74093cf/framework/db/ActiveRecord.php#L471-L474.

arogachev committed:
Add end location to all applicable elements

arogachev commented:
Also could you help with setting up local test environment?

arogachev committed:
Fix testSimpleCreate

arogachev commented:
> Also could you help with setting up local test environment?

Nevermind, I figured this out.

arogachev committed:
Fix remaining tests (unit and integrational)

jaapio commented:
Thanks for this nice improvement! the unittest coverage is below the expected level, that's why the build is now failing.

If you have time to add the required tests that would be very nice, however seen the amount of work, and uncovered code I would accept a decreased coverage for now. :thinking: 

_Side note:
 I think we can improve the tests a bit by using an abstract testcase for our models so we can test the common methods, like `getLocation` and `getEndLocation` in a more flexible way._

jaapio commented:
After merging the first PR, we to have some merge conflicts here. Could you please rebase your branch?

arogachev committed:
Merge branch '5.x' into 224-end-line

# Conflicts:
#	src/phpDocumentor/Reflection/Php/Factory/Function_.php
#	src/phpDocumentor/Reflection/Php/Factory/Method.php
#	src/phpDocumentor/Reflection/Php/Function_.php
#	src/phpDocumentor/Reflection/Php/Method.php

arogachev commented:
> After merging the first PR, we to have some merge conflicts here. Could you please rebase your branch?

Sure. Resolved merged conflicts.

arogachev committed:
Fix tests after changing default values for hasReturnByReference

arogachev committed:
Improve test coverage, add abstract test case for elements

arogachev commented:
Again, thanks for a quick response.

I added more checks for `getEndLocation()`, this improved test coverage, but it's still slightly below the needed level. Maybe it has something to do with `@covers` / `@inheritdoc` tags, not really sure... Also added separate abstract test case for elements that you suggested. 

What else you think needs to be tested? Or maybe it's acceptable for now?

arogachev committed:
Add more tests (getHasReturnByReference())

arogachev committed:
(Sniffer) Remove bare "@inheritdoc" entries

arogachev committed:
(Sniffer) Remove newline and underscore prefix

arogachev committed:
Fix phpstan errors

jaapio commented:
Thanks for the update! this looks good.

The coverage isn't much of an issue for now. We can always improve on this.

arogachev commented:
Ah, you beat me with merging. :slightly_smiling_face: I added tests for previous PR and it resolved coverage problem. Then fixed almost all new CI errors. Trying to fix the last 2 codestyle errors now

Maybe you can help? Can't figure out this one yet.

- https://github.com/phpDocumentor/Reflection/runs/4657674557?check_suite_focus=true
- https://github.com/slevomat/coding-standard#slevomatcodingstandardcommentingdoccommentspacing-
- https://github.com/slevomat/coding-standard/blob/1ace9db4f34d34692f5342f6e51fe7145472cb7b/SlevomatCodingStandard/Sniffs/Commenting/DocCommentSpacingSniff.php#L523-L531

We can fix this in a separate PR though.

Thanks! This was a last blocker for https://github.com/yiisoft/yii2-apidoc/issues/263. :tada:

arogachev commented:
By the way could you make a new release with these changes? Would be nice instead of binding to commit in dependency.

arogachev commented:
Ah, you fixed it already. :+1: Newline was causing the error, got it.

jaapio commented:
I can create that release, later this week. 

It would be nice to have the last part regarding the pretty printer in, so we have a nice new package of features :-). Do you have time to create a last pr for that one?

arogachev commented:
OK. I'll try. :ok_hand:
